patsy/util.py

```diff
@@ -45,18 +45,13 @@
     _pandas_is_categorical_dtype = None
 else:
     if hasattr(pandas, "api"):
```

Comment:
I think you could start with `hasattr(pandas, "CategoricalDType")` and then the lambda

Failing this, you could do `elif hasattr(pandas, "api")`, and then `_pandas_is_categorical_dtype = getattr(pandas.api.types, "is_categorical_dtype", None)`, 

finally, else, `None`.

Something like 
```python
if hasattr(pandas, "CategoricalDType"):
    _pandas_is_categorical_dtype = lambda x: isinstance(x, pandas.CategoricalDType)
elif hasattr(pandas, "api"):
    if getattr(pandas.api.types, "is_categorical_dtype", None) is not None:
        def wrapped_is_categorigal_type(typ):
            with warnings.catch_warnings():
                warnings.filterwarnings('ignore')
                return pandas.api.types.is_categorical_dtype(typ)

        _pandas_is_categorical_dtype = wrapped_is_categorigal_type
else: 
    _pandas_is_categorical_dtype  = None
```

Reply:
Yeah... still fiddling with this a bit... delving into pandas ancient history and seeing where things were introduced.

Reply:
Why do we need the warnings filter?

Reply:
I reckon just throw it in to make sure noting comes up.  We know if is there then it will either FutureWarn, DepreactionWarn, or nothing. The first two are just noise.

Reply:
The current code looks like:
```
have_pandas_categorical = (have_pandas and hasattr(pandas, "Categorical"))
if not have_pandas:
    _pandas_is_categorical_dtype = None
else:
    if hasattr(pandas, "CategoricalDtype"):  # pandas >= 0.25
        _pandas_is_categorical_dtype = lambda x: isinstance(getattr(x, "dtype", x), pandas.CategoricalDtype)
    elif hasattr(pandas, "api"):  # pandas >= 0.19
        _pandas_is_categorical_dtype = getattr(pandas.api.types, "is_categorical_dtype", None)
    else:  # pandas <=0.18
        _pandas_is_categorical_dtype = getattr(pandas.core.common,
                                               "is_categorical_dtype", None)
have_pandas_categorical_dtype = _pandas_is_categorical_dtype is not None
```

I don't see how we could ever generate a deprecation warning... so I'll just drop the warnings filter for now. It's arcane enough as it is :).